NeuroID/src/main/java/com/neuroid/tracker/service/NIDNetworkListener.kt

```diff
@@ -18,14 +20,63 @@ import java.util.Calendar
  * Listening to these messages requires android.permission.ACCESS_NETWORK_STATE set in the
  * manifest. Anytime we get network changes, we send these back as NETWORK_STATE events and
  * update the isConnected flag in NeuroID.
+ *
+ * In event of "not having a network connection" change, the listener will do the following action:
+ * cancel any currently running pause or resume event collection jobs
+ * if event collection is not running, it will take no action
+ * if event collection is running, it will wait 10 seconds and pause event collection
+ *
+ * In event of "having a network connection" change, the listener will do the following action:
+ * cancel any currently running pause or resume event collection jobs
+ * if event collection is running, it will take no action
+ * if event collection is not running, it will wait 10 seconds and resume event collection
  */
 class NIDNetworkListener(private val connectivityManager: ConnectivityManager,
                          private val dataStoreManager: NIDDataStoreManager,
-                         private val neuroID: NeuroID): BroadcastReceiver() {
+                         private val neuroID: NeuroID,
+                         private val dispatcher: CoroutineContext,
+                         private val sleepInterval: Long = SLEEP_INTERVAL): BroadcastReceiver() {
+
+    companion object {
+        const val SLEEP_INTERVAL = 10000L
+    }
+
+    private var haveNoNetworkJob: Job? = null
+    private var haveNetworkJob: Job? = null
+
     override fun onReceive(context: Context?, intent: Intent?) {
         intent?.let {
             if (ConnectivityManager.CONNECTIVITY_ACTION == it.action) {
                 neuroID.isConnected = onNetworkAction()
+
+                // act on the network change
+                handleNetworkChange()
+            }
+        }
+    }
+
+    private fun cancelJobs() {
+        haveNetworkJob?.cancel()
+        haveNoNetworkJob?.cancel()
+    }
+
+    private fun handleNetworkChange() {
+        cancelJobs()
+        if (!neuroID.isConnected) {
+            if (neuroID.isStopped()) {
+                return
+            }
+            haveNoNetworkJob = CoroutineScope(dispatcher).launch {
+                delay(sleepInterval)
+                neuroID.pauseCollection(false)
+            }
+        } else {
+            if (!neuroID.isStopped()) {
+                return
+            }
+            haveNetworkJob = CoroutineScope(dispatcher).launch {
+                delay(sleepInterval)
+                neuroID.resumeCollection()
```

Comment:
I think we need to add another condition here because the SDK could be stopped because they called stopSession and then the network disconnects and reconnects in which case we don't want to start collecting. In iOS we check if there is a connection, the SDK is NOT running, AND there is a userID already set which would indicate our session should still be able to continue

Reply:
Right, good catch! Thanks. Will add that in there.

Reply:
Ah I see, right. Good point. I'll put in a check to see if the userID is set and if not, simply return.